Hermes/appscale/hermes/stats/pubsub_base.py

```diff
@@ -0,0 +1,147 @@
+""" Base classes for building stats publishers and subscribers """
+import logging
+import sys
+
+from tornado.ioloop import PeriodicCallback, IOLoop
+
+
+class SubscriberIsAlreadyRegistered(Exception):
+  """ Raised when you register single subscriber twice on one publisher """
+  pass
+
+
+class SubscriberIsNotRegistered(Exception):
+  """ Raised when you unregister subscriber which is not registered yet """
+  pass
+
+
+class StatsSource(object):
+  """
+  Base class for producers of any kind of stats.
+  Subclasses should implement method get_current()
+  """
+  @property
+  def stats_name(self):
+    return self.__class__.__name__
+
+  def __repr__(self):
+    return self.stats_name
+
+  def get_current(self):
+    """ Returns: current value of specific kind of stats """
+    raise NotImplementedError()
+
+
+class AsyncStatsSource(object):
+  """
+  Base class for asynchronous producers of any kind of stats.
+  Subclasses should implement get_current_async()
+  """
+  @property
+  def stats_name(self):
+    return self.__class__.__name__
+
+  def __repr__(self):
+    return self.stats_name
+
+  def get_current_async(self):
+    """ Returns: Future wrapper for current value of specific kind of stats """
+    raise NotImplementedError()
+
+
+class StatsSubscriber(object):
+  """
+  Base class for stats consumers.
+  Subclasses should implement method receive(stats)
+  """
+  @property
+  def subscriber_name(self):
+      return self.__class__.__name__
+
+  def __repr__(self):
+    return self.subscriber_name
+
+  def receive(self, stats):
+    """ Handlers produced stats entity
+    Args:
+      stats: an object containing stats of some kind (node, processes, ...)
+    """
+    raise NotImplementedError()
+
+
+class StatsPublisher(object):
```

Comment:
I don't feel strongly about it, - your arguments make sense. But how do you suggest to change this?
I have object which can produce stats and I known that produced stats should be put into cache and sent to portal and profile log.
Do you think it's better to remove subscribers list from stats publisher and replace it with 3 fixed attributes: cache, profile_log, and portal? Then we may want to replace single StatsPublisher class with two: LocalStatsUpdater (updates only local cache) and ClusterStatsUpdater (updates local cache and sends collected stats to portal and writes it to profile log).

Yeah.. sounds fine. But I'm not sure if it worth to do. @cdonati , would you suggest to reimplement it this way?